File: 프로그래머스/2/12952. N－Queen/N－Queen.java
import java.util.*;

class Solution {
    int answer = 0;
    int[] col;  // col[row] = row행의 열 값
    public int solution(int n) {        
        col = new int[n];        
        bt(n, 0);
        return answer;
    }
    
    public void bt(int n, int row){
        if(row == n){
            answer++;
            return;
        }
        
        for(int i = 0; i < n; i++){
            col[row] = i;
            if (isPossible(row)) {
                bt(n, row + 1);
            }
        }
    }
    
    public boolean isPossible(int row){
        for(int i = 0; i < row; i++){
            if(col[i] == col[row]) return false;            
            if(Math.abs(row - i) == Math.abs(col[row] - col[i])) return false;
        }
        return true;
    }
}

File: 프로그래머스/2/12952. N－Queen/N－Queen.py
def solution(n):
    answer = 0
    # 체크 배열: 해당 영역에 이미 퀸이 있는지 여부
    check_col = [False] * n
    check_diag1 = [False] * (2 * n) # / 방향 (row + col)
    check_diag2 = [False] * (2 * n) # \ 방향 (row - col + n)

    def bt(row):
        nonlocal answer
        if row == n:
            answer += 1
            return

        for col in range(n):
            if not check_col[col] and not check_diag1[row + col] and not check_diag2[row - col + n]:
                check_col[col] = check_diag1[row + col] = check_diag2[row - col + n] = True
                bt(row + 1)
                check_col[col] = check_diag1[row + col] = check_diag2[row - col + n] = False

    bt(0)
    return answer